File: 503-next-greater-element-ii/next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {

        int n = nums.size();
        vector<int>answer(n,-1);
        stack<int>st;

        for(int i=0;i<2*n;i++)
        {
            while(!st.empty() && nums[i % n] > nums[st.top() % n])
            {
                answer[st.top() % n] = nums[i%n];
                st.pop();
            }

            if(i<n)st.push(i);
        }
        return answer;
        
    }
};

File: 503-next-greater-element-ii/next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] answer = new int[n];
        Stack<Integer> st = new Stack<>();

        // Initialize all answers to -1
        Arrays.fill(answer, -1);

        // Loop twice (to simulate circular array)
        for (int i = 0; i < 2 * n; i++) {
            int curr = nums[i % n];

            // While stack not empty and current element is greater
            while (!st.isEmpty() && nums[st.peek()] < curr) {
                answer[st.pop()] = curr;
            }

            // Only push indices from the first pass
            if (i < n) {
                st.push(i);
            }
        }

        return answer;
    }
}
